lib/perso.ml

```diff
@@ -1216,6 +1239,295 @@ let warning_use_has_parents_before_parent (fname, bp, ep) var r =
+                | _ -> true
+              in
+              if birth && death then loop (b.year :: d.year :: acc) l
+              else if birth && not death then loop (b.year :: acc) l
+              else if (not birth) && death then loop (d.year :: acc) l
+              else loop acc l
+          | Some (Dgreg (b, _)), _ -> (
+              match b.prec with
+              | After | Before | About | Maybe | OrYear _ | YearInt _ ->
+                  if not_dead then loop (conf.today.year :: acc) l
+                  else loop acc l
+              | _ ->
+                  if not_dead then loop (conf.today.year :: b.year :: acc) l
+                  else loop (b.year :: acc) l)
+          | _, Some (Dgreg (d, _)) -> (
+              match d.prec with
+              | After | Before | About | Maybe | OrYear _ | YearInt _ ->
+                  loop acc l
+              | _ -> loop (d.year :: acc) l)
+          | _, _ ->
+              if not_dead then loop (conf.today.year :: acc) l else loop acc l)
+    in
+    loop [] l
+  in
+  List.sort_uniq compare dates
+
+let rec ascendants base acc l i =
+  match l with
+  | [] -> acc
+  | (ip, _, _, lev) :: l -> (
+      match get_parents (poi base ip) with
+      | Some ifam ->
+          let cpl = foi base ifam in
+          let ifath = get_father cpl in
+          let imoth = get_mother cpl in
+          let acc = [ (ifath, [], ifath, [ List.hd lev + 1 ]) ] @ acc in
+          let acc = [ (imoth, [], imoth, [ List.hd lev + 1 ]) ] @ acc in
+          ascendants base acc l i
+      | None -> ascendants base acc l i)
+
+(* descendants des ip de liste1 sauf ceux présents dans liste2 *)
+let descendants_aux base liste1 liste2 =
+  let liste2 = List.map (fun (ip, _, _, _) -> ip) liste2 in
+  let rec loop0 acc = function
+    | (ip, ifaml, ipar0, lev) :: l ->
+        let fams = Array.to_list (get_family (poi base ip)) in
+        let chlds =
+          (* accumuler tous les enfants de ip *)
+          let rec loop1 acc fams =
+            (* iterer sur chaque famille *)
+            match fams with
+            | [] -> acc
+            | ifam :: fams ->
+                let children =
+                  let rec loop2 acc2 children =
+                    match children with
+                    | [] -> acc2
+                    | ipch :: children ->
+                        loop2
+                          ((ipch, ifam :: ifaml, ipar0, [ List.hd lev - 1 ])
+                          :: acc2)
+                          children
+                  in
+                  loop2 [] (Array.to_list (get_children (foi base ifam)))
+                in
+                loop1 (acc @ children) fams
+          in
+          loop1 [] fams
+        in
+        let chlds =
+          List.fold_left (* on élimine les enfants présents dans l2 *)
+            (fun acc (ip, ifaml, ipar, lev) ->
+              if List.mem ip liste2 then acc else (ip, ifaml, ipar, lev) :: acc)
```

Comment:
`List.mem` is not very efficient, if `liste2` can be large it should be a Set instead